File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
import sys
input = sys.stdin.readline
n, k = map(int, input().split())
obj = [list(map(int, input().split())) for _ in range(n)]
dp = [[0 for _ in range(k+1)] for _ in range(n+1)]
for i in range(1, n+1):
    for j in range(1, k+1):
        weight, value = obj[i-1]
        if j >= weight:
            dp[i][j] = max(dp[i-1][j-weight]+value, dp[i-1][j])
        else:
            dp[i][j] = dp[i-1][j]
print(dp[-1][-1])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.cc
#include <iostream>
#include <algorithm>

using namespace std;

int n, k;
int weight[101];
int value[101];
int dp[100001];

int main() {
	cin >> n >> k;
	for (int i = 1; i <= n; i++)
		cin >> weight[i] >> value[i];
	for (int i = 1; i <= n; i++) {
		for (int j = k; j >= 1; j--) {
			if (j >= weight[i])
				dp[j] = max(dp[j], dp[j-weight[i]] + value[i]);
		}
	}
	cout << dp[k];
}
